src/script/printer/relax/utils.h

```diff
@@ -141,7 +141,7 @@ inline int FindVDeviceIndexByTargetKind(const VDevice& vdevice, const IRDocsifie
   int kind_index = 0;
   for (size_t i = 0; i < vdevices.size(); ++i) {
     auto vdev = Downcast<VDevice>(vdevices[i]);
-    if (vdev.same_as(vdevice)) {
+    if (vdev == vdevice) {
```

Comment:
ideally we would like to deduplicate vdevice globally via ptr equality, is it possible to get the pass to generate the same vdevice instead?

Reply:
Should be possible (I think RealizeVDevice pass does similar). Let me explore...

Reply:
@tqchen have made changes to maintain ptr equality for the pass (AnnotateCustomMemoryScope). But, we have an issue as detailed below.

The vdevices GlobalInfo is populated from the pass (AnnotateCustomMemoryScope). Now, when the mod is printed in python after this pass the ptr equality fails, i.e vdevice id printed as -1 ( ``vdevice="opencl:-1:global"``). Additionally, printing mod just before returning ```AnnotateCustomMemoryScope``` or  while in next pass (cpp) works fine.

Also, if the vdevices are populated from python (text cases) and accessed in cpp seem to be working fine.

Functionally, I can proceed here as further pipeline is in cpp. Pls advice.

Reply:
> Seems this is a bug in our parsing mechanism, I think we should fix the parsing in such case to remap things back to ptr equality. if there is a min example, we can work together to fix. cc @yongwww

This is fine now. It was my fault where I tried to print only the ``main`` function from mod and expect appropriate ``VDevice`` index. We are good here.